### PR TITLE
Add proper handling of integer-overflows

### PR DESCRIPTION
**Pull request by jerhard**
In order to implement proper handling of integer overflows in the analyzed C code, and to be able to deal with values outside the range of OCaml-`int64` (#120), we have to make several changes to the integer domains.

1. Add `ikind` information to the integer domains.
2. Use arbitrary precision integers in the implementation of the domains.
3. In the implementation of the abstract operations of the integer domains, add detection & handling of overflows.

Regarding 1., I added an functor `IntDomLifter`, that takes a module of signature `IntDomain.S` and adds a layer to store the ikind. 

TODOs:
- it is not clear how the `top` and `bot` should look like in the `IntDomLifter`. These either would have to be of some arbitrary `ikind` or some additional `Top` and `Bottom` element 
- 2 and 3. still have to be implemented

### CODE REVIEW

**Comment by vogler** [comment]
Instead of copying `module type S` to `module type Z`, it'd be better to just include the common values.

**Comment by vogler** [comment]
Instead of carrying `ikind` around we could add it as an argument for operations. Adding an indirection and additional 64 bit for `ikind` for every abstract int might be costly.
Since a lot of bot/top handling is shared, it's maybe a good idea anyway to merge operations into an `unop`/`binop` which gets the operation and `ikind`.
If we still want to keep the specialized functions like `add`, `sub` etc., they could be added by some functor.

File: src/cdomains/valueDomain.ml [hunk 1]
```diff
@@ -288,7 +288,7 @@ struct
         | TInt (ik,_) ->
           `Int (ID.cast_to ?torg ik (match v with
               | `Int x -> x
-              | `Address x when AD.equal x AD.null_ptr -> ID.cast_to ?torg (ptr_ikind ()) @@ ID.of_int Int64.zero
+              | `Address x when AD.equal x AD.null_ptr -> ID.cast_to ?torg (ptr_ikind ()) @@ ID.of_int (Cilfacade.ptrdiff_ikind ()) Int64.zero
```
Comment (michael-schwarz) [inline src/cdomains/valueDomain.ml:291]: What is the reason behind making a `0` of `Cilfacade.ptrdiff_ikind ()` and then casting it to `ptr_ikind ()`?

File: src/cdomains/valueDomain.ml [hunk 2]
```diff
@@ -389,8 +389,9 @@ struct
         | _ -> AD.join y AD.top_ptr)
     | (`Address x, `Address y) -> `Address (AD.join x y)
     | (`Struct x, `Struct y) -> `Struct (Structs.join x y)
-    | (`Union x, `Union y) -> `Union (Unions.join x y)
-    | (`Array x, `Array y) -> `Array (CArrays.join x y)
+    | (`Union (f,x), `Union (g,y)) -> `Union (match UnionDomain.Field.join f g with
+        | `Lifted f -> (`Lifted f, join x y) (* f = g *)
+        | x -> (x, `Top)) (* f <> g *)    | (`Array x, `Array y) -> `Array (CArrays.join x y)
```
Comment (michael-schwarz) [inline src/cdomains/valueDomain.ml:394]: Move `| (Array x, Array y) -> Array (CArrays.join x y)` down one line again?

**Comment by jerhard** [comment]
>Instead of carrying ikind around we could add it as an argument for operations.
I think there are places besides the arithmetic operations where we want to know the `ikind`. E.g. if we perform a `widen` on ints, it would be useful to know what the `ikind` of the integers is that is being joined.

**Comment by vogler** [comment]
> > Instead of carrying ikind around we could add it as an argument for operations.
> 
> I think there are places besides the arithmetic operations where we want to know the `ikind`. E.g. if we perform a `widen` on ints, it would be useful to know what the `ikind` of the integers is that is being widened.

Might be worthwhile to think about whether this would actually make a difference.
After `widen` an interval might become bigger than the range of its type, but this can then be recovered by first normalizing arguments in any operation. The only case where it could surface is widened values that are not operated on afterwards. However, those we could normalize to their type's range after the solver is done.

File: src/analyses/base.ml.orig [hunk 3]
```diff
@@ -2172,32 +2176,31 @@ struct
...
-                                  (eval_lv ctx.ask gs st lv, `Address (AD.from_var_offset (heap_var, `Index (IdxDom.of_int (Cilfacade.ptrdiff_ikind ()) BI.zero, `NoOffset))))]
-=======
           let heap_var = heap_var ctx in
           let add_null addr =
             if get_bool "exp.malloc.fail"
             then AD.join addr AD.null_ptr (* calloc can fail and return NULL *)
             else addr in
-          set_many ctx.ask gs st [(add_null (AD.from_var heap_var), `Array (CArrays.make (IdxDom.of_int Int64.one) (`Blob (VD.bot (), eval_int ctx.ask gs st size)))); (* TODO why? should be zero-initialized *)
-                                  (eval_lv ctx.ask gs st lv, `Address (add_null (AD.from_var_offset (heap_var, `Index (IdxDom.of_int 0L, `NoOffset)))))]
+<<<<<<< HEAD
```
Comment (michael-schwarz) [inline src/analyses/base.ml.orig:2191]: this looks a bit like a merge conflict ;)

Comment (jerhard) [reply src/analyses/base.ml.orig]: Ups. I removed the `.orig` files now.

**Comment by michael-schwarz** [comment]
This causes `sv-comp/sv-benchmarks/c/loops/linear_search.c` to unsoundly claim that the error is not reached.

**Comment by michael-schwarz** [comment]
I pushed a problematic example where the ikinds don't line up:

~~~C
    signed char* s = malloc(10*sizeof(signed char));
    s[0] = -5;

    unsigned char* us = s;
    *us = 12;
~~~

This is valid C, and GCC also doesn't complain, but we currently fail with `- Fatal error: exception Failure("ikinds signed char and unsigned char are incompatible. Values: (-5) and (12)")`

This is this issue we discussed where it would be necessary to go to `supertop` for `Int` of different `ikind`.

**Pull request merged**